controllers/stardoguser_controller.go

```diff
@@ -18,36 +18,170 @@ package controllers
+			r.Log.Info("StardogUser not found, ignoring reconcile.", "StardogUser", namespace)
+			return ctrl.Result{Requeue: false}, nil
+		}
+		r.Log.Error(err, "Could not retrieve StardogUser.", "StardogUser", namespace)
+		return ctrl.Result{Requeue: true, RequeueAfter: r.ReconcileInterval}, err
+	}
 
-	return ctrl.Result{}, nil
+	rc := &ReconciliationContext{
+		context:       ctx,
+		conditions:    make(map[StardogConditionType]StardogCondition),
+		stardogClient: stardogrest.NewExtendedBaseClient(),
+		namespace:     namespace.Namespace,
+	}
+	sur := &StardogUserReconciliation{
+		reconciliationContext: rc,
+		resource:              stardogUser,
+	}
+
+	if r.validateSpecification(sur.resource.Spec) != nil {
+		rc.SetStatusCondition(createStatusConditionInvalid(err))
+		rc.SetStatusCondition(createStatusConditionReady(false, "Specification cannot be validated"))
+		return ctrl.Result{}, r.updateStatus(sur)
+	}
+	rc.SetStatusIfExisting(StardogInvalid, v1.ConditionFalse)
+
+	if r.syncUser(sur) != nil {
+		rc.SetStatusCondition(createStatusConditionErrored(err))
+		rc.SetStatusCondition(createStatusConditionReady(false, "Synchronization failed"))
+		return ctrl.Result{}, r.updateStatus(sur)
+	}
+
+	rc.SetStatusCondition(createStatusConditionReady(true, "Synchronized"))
+
+	return ctrl.Result{}, r.updateStatus(sur)
 }
 
 func (r *StardogUserReconciler) SetupWithManager(mgr ctrl.Manager) error {
 	return ctrl.NewControllerManagedBy(mgr).
-		For(&stardogv1alpha1.StardogUser{}).
+		For(&StardogUser{}).
 		Complete(r)
 }
+
+func (r *StardogUserReconciler) validateSpecification(spec StardogUserSpec) error {
+	r.Log.V(1).Info("validating StardogUserSpec")
+	if spec.Credentials.SecretRef == "" {
+		return fmt.Errorf(".spec.Credentials.SecretRef is required")
+	}
+	if spec.StardogInstanceRef == "" {
+		return fmt.Errorf(".spec.StardogInstanceRef is required")
+	}
+	return nil
+}
+
+func (r *StardogUserReconciler) syncUser(sur *StardogUserReconciliation) error {
+	spec := sur.resource.Spec
+	userCredentials := spec.Credentials
+	ctx := sur.reconciliationContext.context
+	namespace := sur.reconciliationContext.namespace
+	stardogInstanceRef := spec.StardogInstanceRef
+	stardogInstance := &StardogInstance{}
+
+	r.Log.V(1).Info("checking StardogInstanceRef in the cluster", "ref", stardogInstanceRef)
+	err := r.Client.Get(ctx, types.NamespacedName{Namespace: namespace, Name: stardogInstanceRef}, stardogInstance)
+	if err != nil {
+		return fmt.Errorf("cannot retrieve spec.StardogInstanceRef %s/%s: %v", namespace, stardogInstanceRef, err)
+	}
+
+	adminCredentials := stardogInstance.Spec.AdminCredentials
+	r.Log.V(1).Info("retrieving admin credentials from Secret", "secret", adminCredentials.Namespace+"/"+adminCredentials.SecretRef)
+	adminUsername, adminPassword, err := getCredentials(ctx, r.Client, adminCredentials)
+	if err != nil {
+		return err
+	}
+
+	if userCredentials.Namespace == "" {
+		userCredentials.Namespace = sur.reconciliationContext.namespace
+	}
```

Comment:
I would avoid changing the spec if not strictly necessary. If you do a Update(spec), then the namespace will now be defined to the current namespace. That makes the spec less portable if you'd do `k get stardoguser -o yaml` and try to import somewhere else, because the output will now contain the namespace of the secret it was first created in. In that case I'd rather implement a `GetNamespace(default string)` getter that returns the given default if the credential's namespace is empty.

Reply:
> I would avoid changing the spec if not strictly necessary. If you do a Update(spec), then the namespace will now be defined to the current namespace. That makes the spec less portable if you'd do `k get stardoguser -o yaml` and try to import somewhere else, because the output will now contain the namespace of the secret it was first created in. In that case I'd rather implement a `GetNamespace(default string)` getter that returns the given default if the credential's namespace is empty.

Well I don't see any issue with that. Flux for example updates it's CRDs with defaults. If you don't specify the namespace then it should explicitly be shown that it take the CRD's namespace. Anyway I don't do any update on the CRD so it shouldn't even be a problem

Reply:
Yes you do: When you add finalizers it will update the whole object.
flux can only do that if there's a default specified in the CRD definition. But how should Flux (or any other operators) know the programmatic defaults?
We have already documented that it defaults to the object's namespace in the documentation. Any post-creation patches doesn't make it explicit at creation-time :)